base/broadcast.jl

```diff
@@ -1244,22 +1225,21 @@ macro __dot__(x)
     esc(__dot__(x))
 end
 
-function Base.show(io::IO, bc::Broadcasted)
-    print(io, "Broadcasted(", bc.f)
-    args = bc.args
-    while args != TupleLLEnd()
-        print(io, ", ", args.head)
-        args = args.rest
-    end
-    print(io, ')')
-end
+Base.show(io::IO, bc::Broadcasted) = print(io, "Broadcasted(", bc.f, ", ", bc.args, ')')
 
-function make_kwsyntax(f, args...; kwargs...)
-    args′ = make_TupleLL(args...)
-    g = (args...)->f(args...; kwargs...)
-    return Broadcasted(g, args′)
+struct TypeArgFunction{F, T, n} <: Function
+    f::F
 end
-make(f, args...) = make(combine_styles(args...), f, args...)
-make(::S, f, args...) where S<:BroadcastStyle = Broadcasted{S}(f, make_TupleLL(args...))
+(f::TypeArgFunction{F, T, 1})(args...) where {F,T} = f.f(T, args...)
+(f::TypeArgFunction{F, T, 2})(arg1, args...) where {F,T} = f.f(arg1, T, args...)
+
+@inline make_kwsyntax(f, args...; kwargs...) = make((args...)->f(args...; kwargs...), args...)
+@inline make(f, args...) = _make(f, args...)
+# We use an internal function to hoist out some type information to prevent it
+# from getting lost as a DataType in a tuple.
+@inline _make(f, ::Type{T}, args...) where {T} = make(combine_styles(args...), TypeArgFunction{Core.typeof(f), T, 1}(f), args...)
+@inline _make(f, arg1, ::Type{T}, args...) where {T} = make(combine_styles(arg1, args...), TypeArgFunction{Core.typeof(f), T, 2}(f), arg1, args...)
+@inline _make(f, args...) = make(combine_styles(args...), f, args...)
+@inline make(::S, f, args...) where S<:BroadcastStyle = Broadcasted{S}(f, args)
```

Comment:
`where {S <: BroadcastStyle}`

Reply:
Also, this wasn't introduced by this PR, but the function signatures for `make` conflict here. While it's not common to try to broadcast a `BroadcastStyle` call, it shouldn't be doing type-based dispatch on the `f::Function` argument:
```
+@inline make(f, args...) = _make(f, args...)
```
```
+@inline make(::S, f, args...) where S<:BroadcastStyle = Broadcasted{S}(f, args)
 ```